webapp/src/js/enketo/lib/timer-animation.js

```diff
@@ -0,0 +1,92 @@
+const $ = require('jquery');
+
+const ACTIVE_BG_COLOR = '#0000ff';
+const INACTIVE_BG_COLOR = '#cccccc';
+const ARC_COLOR = '#cccccc';
+
+const drawArc = (ctx, c, arc) => {
+  const arcRadians = Math.PI * arc / 180;
+  ctx.beginPath();
+  ctx.arc(c.x, c.y, c.r, -Math.PI / 2, arcRadians - (Math.PI / 2), false);
+  ctx.lineTo(c.x, c.y);
+  ctx.fillStyle = c.color;
+  ctx.fill();
+};
+
+const drawCircle = (ctx, c) => drawArc(ctx, c, 360);
+
+const drawBackgroundCircle = (ctx, c, color) => drawCircle(ctx, { ...c, color });
+
+const drawTimerArc = (ctx, c, offset, lim) => drawArc(
+  ctx,
+  { ...c, color: ARC_COLOR },
+  offset * 180 / lim
+);
+
+const drawAnimation = (ctx, circle, offset, lim) => {
+  drawBackgroundCircle(ctx, circle, ACTIVE_BG_COLOR);
+  drawTimerArc(ctx, circle, offset, lim);
+};
+
+const setupAudio = () => {
+  const androidSoundSupport = window.medicmobile_android &&
+    typeof window.medicmobile_android.playAlert === 'function';
+  if (androidSoundSupport) {
+    return { play: () => window.medicmobile_android.playAlert() };
+  }
+  return new Audio('/audio/alert.mp3');
+};
+
+const audio = setupAudio();
+
+const animate = function(canvas, duration, onComplete) {
```

Comment:
So, I spent way longer than I should have agonizing over the best way to structure this code and I am still not at all happy with how it is.  In general I am not a fan of nesting functions in other functions to create a pseudo-class. I think it is terrible for readability and maintainability since everything in the function has access to so much state and the logic flow can jump all over the place.  This case is extra tricky since so much of the logic is tied to the mutable state of `running`.  From a readability perspective, I feel like a proper class here would be better than nested functions. However, JS classes are complete trash until ES2022 when they finally added support for [private properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties) (while we are still targeting ES2018) and I could not find any other place in our entire code base where we actually used a JS class.  

So, in the end I tried to split the difference and break out the functions that did not rely on the internal state of `animate` so they could be considered separately while only keeping the internal functions in `animate` that _required_ access to the internal state of that function....  But like I said, I am still not really happy with how things turned out.  Let me know your thoughts here!  I am happy to just put all the functions back into `animate` if you think that would be best!